### PR TITLE
doc: Add Redeemer docs

### LINKED ISSUE #1532: Ticket Redemption Service docs
Add documentation and sequence diagrams to `go-livepeer/doc` for the Ticket Redemption Service

### PR DESCRIPTION
**Pull request by kyriediculous**
**What does this pull request do? Explain your changes. (required)**
This PR adds documentation for the `Redeemer` service and `ticketQueue` table. 

**Does this pull request close any open issues?**
Fixes #1532 

**Checklist:**
- [x] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

File: doc/database.md [hunk 1]
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
...
+## Table `ticketQueue`
+
+**orchestrator/redeemer** only. `ticketQueue` Tracks winning tickets for probabilistic micropayments.
+
+Column | Type | Description
+---|---|---
+createdAt | DATETIME DEFAULT CURRENT_TIMESTAMP | Time this row was inserted 
+sender | STRING | Address of the broadcaster that sent the winning ticket.
+recipient | STRING | Address of the orchestrator that payments will be credited to.
+faceValue | BLOB | Face value of the ticket, in wei Unique nonce sent by the broadcaster.
```
Comment (yondonfu) [inline doc/database.md]: "Unique nonce sent by the broadcaster" can be removed. I see that this was also in the description for the `winningTickets` table...not sure how that got there 🤷‍♂️

File: doc/database.md [hunk 2]
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
...
+winProb | BLOB | The ticket's winning probability in the range of 0 through 2^256-1.
+senderNonce | INTEGER | Nonce incorporated by the broadcaster with each ticket.
+recipientRand | BLOB | Value used by the orchestrator when constructing the initial ticket parameters.
+recipientRandHash | STRING | Hash of the recipient rand, keccak256(recipientRand).,
```
Comment (yondonfu) [inline doc/database.md]: Let's remove the trailing commas for some of these descriptions.

File: doc/redeemer.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _This will trigger a `LocalSenderMonitor.SubscribeMaxFloatChange(ticket.sender)` notification_
+
+## Monitoring Max Float
+
+1. On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server for `maxFloat` updates from broadcasters the orchestrator (or all orchestrators in a multi-orchestrator setup) is working with.* 
+
+2. Anytime a winning ticket is submitted to the `Redeemer Server`, the `Redeemer Server` will monitor it's own `LocalSenderMonitor` for `maxFloat` change notifications for the sender of the ticket using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` (if not already watching for updates from that particular sender). 
+
+3. Whenever the subscription from `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` receives a notificiation, the `Redeemer Server` will request the latest `maxFloat` for the `sender` using `LocalSenderMonitor.MaxFloat(sender)`.
```
Comment (yondonfu) [inline doc/redeemer.md]: "notificiation" -> "notification"

File: doc/database.md [hunk 4]
```diff
@@ -64,3 +64,24 @@ recipientRand | BLOB | Value used by the orchestrator when constructing the init
 recipientRandHash | STRING | Hash of the recipient rand, keccak256(recipientRand).
 sig | BLOB | The broadcaster's signature over the ticket parameters.
 sessionID | STRING | Broadcast session which this ticket belongs to.
+
+## Table `ticketQueue`
+
+**orchestrator/redeemer** only. `ticketQueue` Tracks winning tickets for probabilistic micropayments.
```
Comment (yondonfu) [inline doc/database.md]: Capitalize orchestrator and redeemer to match style used in the rest of this document (for example, note the "Orchestrator only" phrase for the `winningTickets` table section).

File: doc/redeemer.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+So far we've discussed `LocalSenderMonitor.addFloat()` and `LocalSenderMonitor.subFloat()` being responsible for triggering `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` notifications, but these can also be triggered by certain Ethereum events related to the Livepeer protocol: 
+
+- FundReserve: When a broadcaster funds it's reserve the `maxFloat` allocation increases by the added reserve divided by the active orchestrator set size.
+- NewRound: If the active orchestrator set size changes, the `maxFloat` will become the current broadcaster's reserve divided by the new active set size. Since this event impacts all participants in the protocol the `Redeemer` will have to send updates for _every_ `sender` it is keeping track of. 
+
+![Ethereum Events](https://imgur.com/Gzf4BD0.png)
+
+
+
+## Networking
```
Comment (yondonfu) [inline doc/redeemer.md]: Let's add this section to [doc/networking.md](https://github.com/livepeer/go-livepeer/blob/master/doc/networking.md) following the format used there so that all networking docs are in one place. We can link to the relevant section of the networking docs from this document so that someone reading this document to understand the ticket redemption service flow can easily find networking information for the ticket redemption service.

File: doc/redeemer.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+
+3. Whenever the subscription from `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` receives a notificiation, the `Redeemer Server` will request the latest `maxFloat` for the `sender` using `LocalSenderMonitor.MaxFloat(sender)`. 
+
+4. The `Redeemer Server` will use the return value from [3] to push a `MaxFloatUpdate` to it's subscribed orchestrators. If not already existing, this update will create a local cache of the `maxfloat` value for the `sender`.*
+
+\* In a multi-O setup `MonitorMaxFloatUpdate` is sent to all orchestrators in the cluster and a local cache is kept, even if the orchestrator is not currently working with the broadcaster but another orchestrator in the cluster is. This allows a broadcaster to smoothly fail over to other orchestrators within the same multi-O cluster but also creates additional storage and networking overhead. Unused local cache is cleaned up on a set interval. 
+
+![Ticket Flow](https://i.imgur.com/jd83N3c.png)
```
Comment (yondonfu) [inline doc/redeemer.md]: IIUC, this diagram indicates that the Redeemer only calls SubscribeMaxFloatChange for a B address when the Redeemer queues a winning ticket from that B and the Redeemer is not yet aware of that B. This would mean that an O would not receive max float updates for a B that is sending it payments until O receives a winning ticket from B. In this scenario, if B's max float changes due to a ReserveFunded or NewRound event, O would not receive those updates which impacts the ticket params/pricing that it advertises to B.

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: doc/redeemer.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
+# Ticket Redemption Service
+
+The ticket redemption service allows an Orchestrator to run a separate service for redeeming winning tickets on-chain using a separate ETH account.
```
Comment (j0sh) [inline doc/redeemer.md]: > allows an Orchestrator to run a separate service

It might get confusing with "orchestrator-the-node" versus "orchestrator-the-operator", since redemption services are not going to be orchestrators.

> using a separate ETH account.

Maybe clarify "separate ETH account on the redemption service". 

Altogether, this could be tweaked to something like: "The ticket redemption service allows orchestrator nodes to redeem winning tickets on-chain using a separate ETH account on the ticket redemption service. Orchestrator nodes no longer need to have an account with ETH to pay for transactions."

File: doc/redeemer.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
+# Ticket Redemption Service
+
+The ticket redemption service allows an Orchestrator to run a separate service for redeeming winning tickets on-chain using a separate ETH account. 
+
+The ticket redemption service is responsible for redeeming winning tickets as well as pushing max float updates about broadcasters back to the orchestrator(s).
```
Comment (j0sh) [inline doc/redeemer.md]: Is there a doc anywhere that describes the particulars of the PM mechanism, specifically the max float? It might be helpful to link to that here, so readers understand what a max float is.

**Review by j0sh — CHANGES_REQUESTED** [review]
It would be good to also have usage instructions, maybe after the introduction, with recommended CLI flags for both the redeemer service and the orchestrator node.

File: doc/redeemer.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+The ticket redemption service allows an Orchestrator to run a separate service for redeeming winning tickets on-chain using a separate ETH account. 
+
+The ticket redemption service is responsible for redeeming winning tickets as well as pushing max float updates about broadcasters back to the orchestrator(s). 
+
+## TicketQueue
+
+1. The `ticketQueue` is a loop that runs everytime a new block is seen. It will then pop tickets off the queue starting with the oldest ticket first, and sends it to the `LocalSenderMonitor` for redemption if the `recipientRand` for the ticket has expired. 
+
+2. When the `LocalSenderMonitor` receives a ticket from the `ticketQueue` it will substract `ticket.faceValue` from the outstanding `maxFloat` as long as the ticket is in limbo.
```
Comment (j0sh) [inline doc/redeemer.md:17]: Clarify "in limbo" ? Would it be clearer to say "not yet confirmed on the blockchain" ?

File: doc/redeemer.md (hunk 6 above)
Comment (j0sh) [inline doc/redeemer.md]: If an image is copy-pasted into a github text box, github will upload it automatically to `githubusercontent.com` and present a URL to the image. Using that URL here is probably a bit better than cross-linking to imgur. For example, here's the same image hosted on Github: https://user-images.githubusercontent.com/292510/84299949-776ce480-ab06-11ea-9a88-fa052dc15c85.png

File: doc/redeemer.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+4. When the Ethereum node returns a response the `ticket.faceValue` is added to the `maxFloat` again as the ticket is no longer in limbo.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _This will trigger a `LocalSenderMonitor.SubscribeMaxFloatChange(ticket.sender)` notification_
+
+## Monitoring Max Float
+
+1. On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server for `maxFloat` updates from broadcasters the orchestrator (or all orchestrators in a multi-orchestrator setup) is working with.* 
+
+2. Anytime a winning ticket is submitted to the `Redeemer Server`, the `Redeemer Server` will monitor it's own `LocalSenderMonitor` for `maxFloat` change notifications for the sender of the ticket using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)` (if not already watching for updates from that particular sender).
```
Comment (j0sh) [inline doc/redeemer.md]: Can we clarify what a "sender" is here for folks unfamiliar with the code, or use more direct terminology, like "broadcaster" ?

File: doc/redeemer.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+3. The ticket is sent to a remote Ethereum node for redemption
+
+4. When the Ethereum node returns a response the `ticket.faceValue` is added to the `maxFloat` again as the ticket is no longer in limbo.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _This will trigger a `LocalSenderMonitor.SubscribeMaxFloatChange(ticket.sender)` notification_
+
+## Monitoring Max Float
+
+1. On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server for `maxFloat` updates from broadcasters the orchestrator (or all orchestrators in a multi-orchestrator setup) is working with.*
```
Comment (j0sh) [inline doc/redeemer.md]: Would it be clearer to say, 

"On node startup, the orchestrator subscribes to `MonitorMaxFloat` on the gRPC Redeemer server. Each time the redeemer receives a winning ticket, all subscribed orchestrators receive a `maxFloat` update for the broadcaster that generated the winning ticket. All orchestrators receive this update, regardless of whether the orchestrator is working with the broadcaster."

"Within the `Redeemer Server`, float monitoring works in the following way: ..." 

Then points #2-4 can describe the redeemer's local float monitoring in more detail. (Should we also mention how orchestrator-local float monitoring works in regular mode?)

File: doc/redeemer.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+\* In a multi-O setup `MonitorMaxFloatUpdate` is sent to all orchestrators in the cluster and a local cache is kept, even if the orchestrator is not currently working with the broadcaster but another orchestrator in the cluster is. This allows a broadcaster to smoothly fail over to other orchestrators within the same multi-O cluster but also creates additional storage and networking overhead. Unused local cache is cleaned up on a set interval. 
+
+![Ticket Flow](https://i.imgur.com/jd83N3c.png)
+
+## Discovery
+
+During discovery (`GetOrchestratorInfo`), an orchestrator invokes the `MaxFloat` RPC call. Currently it **does not cache** the returned value but uses it to calculate the `TicketParams` for the `GetOrchestratorInfo` response, then discards it. 
+
+This is done to not start a watchdog and keep local cache for each broadcaster that discovers our node.
```
Comment (j0sh) [inline doc/redeemer.md]: "This is done to not start a watchdog and keep local cache..." is a bit hard to parse.  I wonder if the mention of "no caching" above is sufficient, and this line can be omitted? Or maybe go a bit further into the reasons we don't keep a cache. Eg, "A cache is not kept because the `GetOrchestratorInfo` request is only informational, and does not guarantee work will actually come in. Hence, we avoid persisting local state after each request, such as a cache."

File: doc/database.md (hunk 1 above)
Comment (kyriediculous) [reply doc/database.md]: fixed in e5624db9

File: doc/database.md (hunk 2 above)
Comment (kyriediculous) [reply doc/database.md]: fixed in e5624db9

File: doc/redeemer.md (hunk 8 above)
Comment (kyriediculous) [reply doc/redeemer.md]: There is this old open PR but it doesn't mention max float https://github.com/livepeer/go-livepeer/pull/1099/files

File: doc/redeemer.md (hunk 9 above)
Comment (kyriediculous) [reply doc/redeemer.md:17]: pending between the Redeemer and the ethereum node 

This can be either pending on-chain or an HTTP request to the eth node that has some latency to it.

**Comment by kyriediculous** [comment]
this branch has been force pushed with changes in accordance with https://github.com/livepeer/go-livepeer/pull/1513#issuecomment-644416131

File: doc/database.md (hunk 4 above)
Comment (yondonfu) [reply doc/database.md]: Bump

File: doc/networking.md [hunk 13]
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
...
+
+```
+message MaxFloatUpdate {
+    bytes max_float= 2;
+}
+```
+
+### gRPC `MonitorMaxFloat`: `MaxFloatReq` -> stream `MaxFloatUpdate`
+
+`MonitorMaxFloat` is a server-side streming RPC method that is called by an Orchestrator to receive indirect max float updates for a sender. The stream will remain open until either the client or the server closes the stream or connection.
```
Comment (yondonfu) [inline doc/networking.md]: I think we can drop "indirect".

File: doc/networking.md [hunk 14]
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
...
+message MaxFloatUpdate {
+    bytes max_float= 2;
+}
+```
+
+### gRPC `MonitorMaxFloat`: `MaxFloatReq` -> stream `MaxFloatUpdate`
+
+`MonitorMaxFloat` is a server-side streming RPC method that is called by an Orchestrator to receive indirect max float updates for a sender. The stream will remain open until either the client or the server closes the stream or connection. 
+
+The request follows the same format as the `Maxfloat` method:
```
Comment (yondonfu) [inline doc/networking.md]: `Maxfloat` -> `MaxFloat`.

File: doc/networking.md [hunk 15]
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
...
+
+```
+message MaxFloatReq {
+    bytes sender = 1;
+}
+```
+
+The response message also follows the same format but is now _stream_  of `MaxFloatUpdate` messages:
```
Comment (yondonfu) [inline doc/networking.md]: "_stream_" -> `a _stream_`.

File: doc/networking.md [hunk 16]
```diff
@@ -196,3 +196,60 @@ for updating Protocol Buffers messages. The same principles of "add but don't
 modify or remove fields" carry over to gRPC service definitions: new services
 can be added, but names should not be changed, nor should services be removed
 unless the intent is to break backwards compatibility.
+
+
+# Redeemer networking spec
```
Comment (yondonfu) [inline doc/networking.md]: Let's move this section after the "Ping" section so that it is grouped together with the rest of the networking flow sections. A more descriptive name for this section may be "Remote Ticket Redemption".

File: doc/redeemer.md [hunk 17]
```diff
@@ -0,0 +1,56 @@
+# Ticket Redemption Service
+
+The Ticket Redemption Service allows Orchestrator nodes to redeem winning tickets on-chain using a separate ETH account on the Ticket Redemption Service. Orchestrator nodes no longer need to have an account funded with ETH to pay for transactions on the Ethereum network.
+
+It is responsible for redeeming winning tickets as well as pushing _max float_ updates about broadcasters back to its connected Orchestrators.
+
+_Max float_ is the guaranteed value an Orchestrator will be able from a Broadcaster's reserve. It accounts for the current reserve allocation from a Broadcaster to an Orchestrator as well as pending winning ticket redemptions.
```
Comment (yondonfu) [inline doc/redeemer.md]: "able from a Broadcaster's reserve" -> "able to claim from a Broadcaster's reserve"

File: doc/redeemer.md [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+
+## Monitoring Max Float
+
+1. When max float for a `sender` is requested from the `RedeemerClient` but no local cache is available, an (unary) RPC call will be sent to the `RedeemerServer`. 
+
+2. A second RPC call to `MonitorMaxFloat(sender)` will open up a server-side gRPC stream to receive future update. 
+
+_If this call fails the response from step 1 is returned, but not kept in cache to prevent duplicate streams_
+
+3. The `RedeemerServer` go-routine started by the RPC call in step 2 will start a subscription to listen for max float changes from the `LocalSenderMonitor` for the specified `sender` using `LocalSenderMonitor.SubscribeMaxFloatChange(sender)`.
```
Comment (yondonfu) [inline doc/redeemer.md]: "go-routine" -> "goroutine"

File: doc/redeemer.md [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+3. The ticket is sent to a remote Ethereum node for redemption
+
+4. When the Ethereum node returns a response the `ticket.faceValue` is added to the `maxFloat` again as the ticket is no longer in limbo.
+
+&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; _This will trigger a `LocalSenderMonitor.SubscribeMaxFloatChange(ticket.sender)` notification_
+
+## Monitoring Max Float
+
+1. When max float for a `sender` is requested from the `RedeemerClient` but no local cache is available, an (unary) RPC call will be sent to the `RedeemerServer`.
```
Comment (yondonfu) [inline doc/redeemer.md]: The actual implementation is called `Redeemer` so let's replace `RedeemerServer` with `Redeemer`?

File: doc/redeemer.md (hunk 8 above)
Comment (yondonfu) [reply doc/redeemer.md]: The [PM spec](https://github.com/livepeer/wiki/blob/master/spec/streamflow/pm.md#reserve) provides a description of the float concept. We could add link in the redeemer doc for the reader's reference.

File: doc/networking.md (hunk 16 above)
Comment (kyriediculous) [reply doc/networking.md]: But it's not necessarily part of the core livepeer networking protocol, should we keep those separate?

Comment (yondonfu) [reply doc/networking.md]: There are distinct sections for "Broadcaster to Orchestrator" and "Broadcaster to Transcoder" (and there should probably be a "Orchestrator to Transcoder" section added at some point) so adding an additional, but separate section seems reasonable to me. Given the section naming convention, maybe a better section name would be "Orchestrator to Redeemer".

File: doc/redeemer.md (hunk 6 above)
Comment (kyriediculous) [reply doc/redeemer.md]: I added an assets folder instead so that both the exported PNG as well as the source for the sequence diagram can be part of the repo.

File: doc/redeemer.md (hunk 12 above)
Comment (kyriediculous) [reply doc/redeemer.md]: this has been removed

File: doc/redeemer.md (hunk 11 above)
Comment (kyriediculous) [reply doc/redeemer.md]: this has been removed

File: doc/redeemer.md (hunk 10 above)
Comment (kyriediculous) [reply doc/redeemer.md]: yeah I guess the question is whether we want to use the language of the `pm` protocol or the livepeer protocol.

Perhaps we can add a note that currently sender is interchangeable with broadcaster

File: doc/redeemer.md (hunk 7 above)
Comment (kyriediculous) [reply doc/redeemer.md]: This should be fixed

File: doc/database.md (hunk 4 above)
Comment (kyriediculous) [reply doc/database.md]: fixed in 17af7bb

File: doc/networking.md (hunk 16 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in c29f76c

File: doc/networking.md (hunk 15 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in c29f76c

File: doc/networking.md (hunk 14 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in  c29f76c

File: doc/networking.md (hunk 13 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in c29f76c

File: doc/redeemer.md (hunk 17 above)
Comment (kyriediculous) [reply doc/redeemer.md]: fixed in 9f594b4

File: doc/redeemer.md (hunk 18 above)
Comment (kyriediculous) [reply doc/redeemer.md]: fixed in 1110b4d91bfe9ad1edfb4f658bcd6e2c89e03090

File: doc/redeemer.md (hunk 19 above)
Comment (kyriediculous) [reply doc/redeemer.md]: fixed in fixed in 9f594b4

File: doc/redeemer.md (hunk 8 above)
Comment (kyriediculous) [reply doc/redeemer.md]: I added a note that links to that section of the pm spec

File: doc/networking.md [hunk 20]
```diff
@@ -162,6 +162,65 @@ message PingPong {
...
+message MaxFloatReq {
+    bytes sender = 1;
+}
+```
+
+The server will respond with the `MaxFloat` for that `sender` in raw bytes format:
+
+```
+message MaxFloatUpdate {
+    bytes max_float= 2;
```
Comment (yondonfu) [inline doc/networking.md]: Let's update this to match the definition that was updated as a part of this [comment](https://github.com/livepeer/go-livepeer/pull/1513#discussion_r441216728).

File: doc/networking.md [hunk 21]
```diff
@@ -162,6 +162,65 @@ message PingPong {
...
+message MaxFloatReq {
+    bytes sender = 1;
+}
+```
+
+The response message also follows the same format but is now a _stream_  of `MaxFloatUpdate` messages: 
+
+```
+message MaxFloatUpdate {
+    bytes max_float= 2;
```
Comment (yondonfu) [inline doc/networking.md]: Let's update this to match the definition that was updated as a part of this [comment](https://github.com/livepeer/go-livepeer/pull/1513#discussion_r441216728).

File: doc/networking.md [hunk 22]
```diff
@@ -162,6 +162,65 @@ message PingPong {
...
+}
+```
+
 ### Notes
```
Comment (yondonfu) [inline doc/networking.md]: I think these notes are supposed to be a part of the "Ping" section.

File: doc/redeemer.md [hunk 23]
```diff
@@ -0,0 +1,58 @@
...
+
+## Monitoring Max Float
+
+*A more detailed description about `maxFloat` and it's relation to a broadcaster's reserve can be found in the [PM protocol spec](https://github.com/livepeer/wiki/blob/master/spec/streamflow/pm.md#reserve).*
+
+1. When max float for a `sender` is requested from the `RedeemerClient` but no local cache is available, an (unary) RPC call will be sent to the `Redeemer`. 
+
+2. A second RPC call to `MonitorMaxFloat(sender)` will open up a server-side gRPC stream to receive future update. 
+
+_If this call fails the response from step 1 is returned, but not kept in cache to prevent duplicate streams_
```
Comment (yondonfu) [inline doc/redeemer.md]: Is this referring to step 2? If so, why would keeping the result from step 1 in the cache result in duplicate streams? If step 2 fails, wouldn't there be no stream at all?

**Review by yondonfu — CHANGES_REQUESTED** [review]

File: doc/redeemer.md (hunk 23 above)
Comment (kyriediculous) [reply doc/redeemer.md]: Ah yes that's true, this must be an old line that I repurposed. 

I think it should state something like "to prevent stale cache due to not being able to receive further updates"

Comment (kyriediculous) [reply doc/redeemer.md]: fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

File: doc/networking.md (hunk 22 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

File: doc/networking.md (hunk 21 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

File: doc/networking.md (hunk 20 above)
Comment (kyriediculous) [reply doc/networking.md]: fixed in fcb86dd293a40d0a61916d875394afeb32a80dd5

File: doc/redeemer.md (hunk 10 above)
Comment (kyriediculous) [reply doc/redeemer.md]: fixed in 8a6184d1

**Review by yondonfu — COMMENTED** [review]
Changes look good to me. 👍   for a rebase from me as long as comments from other reviewers are addressed.

File: doc/redeemer.md [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+
+5. Upon retrieving the latest max float value for `sender` it will be sent over the server-side gRPC stream.
+
+6. Upon receiving a `MaxFloatUpdate` over the server-side gRPC stream for `sender` it will update its local cache for that `sender` accordingly. 
+
+7. Subsequent calls to `RedeemerClient.MaxFloat(sender)` will return the locally cached value for `sender` as long as it remains available. 
+
+8. The local cache for `sender` will be cleaned up if is not requested for 5 minutes. 
+
+![Ticket Flow](./assets/redeemer/ticketflow.png)
```
Comment (j0sh) [inline doc/redeemer.md:49]: I really like these sequence diagrams, but 200-300 KB is kind of large. I suppose if we're going to be doing a lot more of these types of checkins, we should look at a blob storage extension such as git-lfs (which Github supports).

Comment (kyriediculous) [reply doc/redeemer.md:49]: Do you want to include this in this PR? I think this is the only remaining issue for this PR, right?

Comment (j0sh) [reply doc/redeemer.md:49]: Up to you, but I'm generally mindful of repo growth especially with large binary blobs in Git. Another option: since the source for these diagrams has been checked in and can be reproduced [1], we could upload the images themselves to Github (drag+drop the image in this text box) and link within the docs?

[1]  Maybe add a comment to the source(s) noting which tools / service can interpret the code for these diagrams?

Comment (kyriediculous) [reply doc/redeemer.md:49]: now using git lfs for `.png` files

**Review by yondonfu — APPROVED** [review]
LGTM

**Review by j0sh — APPROVED** [review]
